### ISSUE TITLE
Training helen landmark indexing

### ISSUE THREAD

**Issue by tgd-git** [takuya-takeuchi/FaceRecognitionDotNet#77]
Hi i have trained a model with the helen dataset. Works great btw! I just wonder how to fix the landmark indexing. It is bit chaotic, consitent chaotic, but still:)  ->
![image](https://user-images.githubusercontent.com/2562637/74363790-944d1180-4dcb-11ea-9a8b-540e77de2968.png)

**Comment by takuya-takeuchi** [comment]
This strange indexing is coming from dataset.
You must modify dataset if you dislike dislike it.

But FaceRecognitionDotNet return the group of location per face parts.
Please check https://github.com/takuya-takeuchi/FaceRecognitionDotNet/blob/1d99e43591f6996910ab1a2abc0facbdd1ee24f4/src/FaceRecognitionDotNet/FaceRecognition.cs#L339

**Comment by tgd-git** [comment]
So i was looking at this for two days now and dont get the logic. Could you please elaborate why these indexes are so strange? The index code says they are mapping is for a 194 landmarks helen trained model? No? When you say "You must modify dataset if you dislike it.", what do you mean? 
I am bit lost. If you have a moment please try to clarify the situation and outline what to do. I am happy to work this but very confused right now. I am not familiar with ML and maybe overthinking this. Please give some hints. Thank you.

**Comment by takuya-takeuchi** [comment]
> I just wonder how to fix the landmark indexing.
At first, I thought you dislike discontinuous index (like 0, 1, 106, 117..). Therefore, I said 'You must modify dataset if you dislike it.'.

However, I checked the dataset again and I found that face parts point order is not strange.

![image](https://user-images.githubusercontent.com/6241854/74586218-bd1e1280-5028-11ea-937a-39544e8e52f5.png)

So training program may have any issue but it looks like working well.

**Comment by tgd-git** [comment]
> At first, I thought you dislike discontinuous index (like 0, 1, 106, 117..). Therefore, I said 'You must modify dataset if you dislike it.'.

Yes exactly, i dont like the order. Still dont understand how to modify the dataset. Feels like i am missing a concept here or confuse wording. do you mean the index mapping you linked to in the code? Yes I wonder why you do the mapping in the code like this. 
Should not all the parts have consecutive indexes like chin 0-100, left eye 101-120 and so on. (numbers just estimated). The only part that differs would be the mouth where two parts (upper and lower lip) share some indexes. It seems pretty much like constructing polygons from vertex indexes. I am familiar with this concept.

> However, I checked the dataset again and I found that face parts point order is not strange.

> So training program may have any issue but it looks like working well.

I do not understand what you mean here. So yes i see your screenshot where the indexing seems ok in the beginning of chin part (0,1,2,3..) Why does mine differ (0, 1, 106, 117..)? I was using the vanilla helentraining example you provide. Still confused sorry :)

**Comment by takuya-takeuchi** [comment]
The above image is created from original dataset image rather than result of inference.

So I face same issue you faced.
This image is result of inference from my helen dataset model.
And target image (helen\helen_1\1115234911_1.jpg) equals to the above image.

![image](https://user-images.githubusercontent.com/6241854/74587511-b4ccd400-5036-11ea-8fe5-3545b68fd291.png)

Therefore, this trange index order is not generated randomly but I don't know why.

**Comment by tgd-git** [comment]
Ok i see. just misunderstanding:) Is it not just the index mapping in PredictorModel.Helen?
When I look at the predictor models in the code i see PredictorModel.Large is pretty logical and clear. PredictorModel.Helen on the other side is confusing. 

Why is the chin part not just:

` { FacePart.Chin,         Enumerable.Range(0,100).Select(i => landmarkTuple[i]) },`

Did not count the chin points, 100 is just for example. 
Also in PredictorModel.Helen you do not use `.ToArray() `

See PredictorModel.Large and  PredictorModel.Small uses  `.ToArray() ` at the end of the statement.

`{ FacePart.Chin,         Enumerable.Range(0,17).Select(i => landmarkTuple[i]).ToArray() },`

**Comment by takuya-takeuchi** [comment]
I think your provided image (on first comment) is generated by HelenTraining demo command.
In this demo command, does NOT use FaceRecognition.FaceLandmark method to draw face part number like your image.

This code is the part of HelenTraining demo command process.

````
                    using (var predictor = ShapePredictor.Deserialize(modelFile))
                    using (var image = FaceRecognition.LoadImageFile(imageFile))
                    using (var mat = Dlib.LoadImageAsMatrix<RgbPixel>(imageFile))
                    using (var bitmap = (Bitmap)System.Drawing.Image.FromFile(imageFile))
                    using (var white = new Bitmap(bitmap.Width, bitmap.Height))
                    using (var g = Graphics.FromImage(bitmap))
                    using (var gw = Graphics.FromImage(white))
                    {
                        var loc = _FaceRecognition.FaceLocations(image).FirstOrDefault();
                        if (loc == null)
                        {
                            Console.WriteLine("No face is detected");
                            return 0;
                        }

                        var b = new DlibDotNet.Rectangle(loc.Left, loc.Top, loc.Right, loc.Bottom);
                        var detection = predictor.Detect(mat, b);

                        using (var p = new Pen(Color.Red, bitmap.Width / 200f))
                        {
                            g.DrawRectangle(p, loc.Left, b.Top, b.Width, b.Height);
                            gw.Clear(Color.White);
                            gw.DrawRectangle(p, loc.Left, b.Top, b.Width, b.Height);
                        }

                        for (int i = 0, parts = (int)detection.Parts; i < parts; i++)
                        {
                            var part = detection.GetPart((uint)i);
                            g.FillEllipse(Brushes.GreenYellow, part.X, part.Y, 15, 15);
                            gw.DrawString($"{i}", SystemFonts.DefaultFont, Brushes.Black, part.X, part.Y);
                        }

                        bitmap.Save("demo.jpg", ImageFormat.Jpeg);
                        white.Save("white.jpg", ImageFormat.Jpeg);
````

The part number of your image is drawn from `gw` object.
Please look at `detection` object which comes from `predictor.Detect` rather than FaceRecogniton object.

`predictor` is generated from Deserialzie method of DlibDotNet and source is trained model file.
Yes, this strange index comes from raw result of dlib. Not comes from `FaceRecognition.FaceLandmark` you pointed out on the avove your comment.

**Comment by takuya-takeuchi** [comment]
# Memo

Training parameter may have difference between python anc c++.
* https://github.com/davisking/dlib/blob/master/tools/python/src/shape_predictor.h
* https://github.com/davisking/dlib/blob/master/dlib/image_processing/shape_predictor_trainer_abstract.h

Should we check default parameter?

**Comment by tgd-git** [comment]
did you had a chance to look into it? i am still lost. it is hard for me to understand what is going on and where to look for a fix.

**Comment by takuya-takeuchi** [comment]
I referred this blog.
https://qiita.com/kekeho/items/0b2d4ed5192a4c90a0ac

This blog uses python but it works fine.
So it may help us.

**Comment by tgd-git** [comment]
The example from the blog has the same issue. There is a "fix" script in the comments that remaps the indexes. That is also what i had already in mind but it feels pretty dirty.
The last comment gave me bit hope though. It states that the issue can be resolved by padding the name attribute with zeros (trailing zeros). 
I tried this and have generated xml data like this:

`<part x="565" y="758" name="000" /><part x="564" y="781" name="001" />` and so on..

Following this i trained again and run the HelenTraining demo code but no luck :(
The issue persists  :(

**Comment by takuya-takeuchi** [comment]
name attribute could be first item.
https://github.com/davisking/dlib/blob/fc6992ac0422acdaba77cdb07678605ca60f6321/dlib/data_io/image_dataset_metadata.cpp#L105

and patrs is std::map and it could be sorted by key.
https://github.com/davisking/dlib/blob/fc6992ac0422acdaba77cdb07678605ca60f6321/dlib/data_io/image_dataset_metadata.h#L70

so I thought your fix should resolve but isuue remain.

**Comment by takuya-takeuchi** [comment]
@tgd-git 
I added 0 padding to name attribute and it works fine.
And I used latest DlibDotNet but it should affect nothing.

````diff
- parts.Add(new Part { X = tmp[0], Y = tmp[1], Name = $"{i - 1}" });
+ parts.Add(new Part { X = tmp[0], Y = tmp[1], Name = $"{i - 1:D3}" });
````

````
dotnet run -c Release -- demo --model helen\helen-dataset.dat --directory models --image helen\helen_1\100040721_2.jpg
````

![white](https://user-images.githubusercontent.com/6241854/75362576-48868600-58fc-11ea-9a05-d6ea2988b6d9.jpg)
![demo](https://user-images.githubusercontent.com/6241854/75362578-491f1c80-58fc-11ea-9827-220f4b397467.jpg)

> The last comment gave me bit hope though. It states that the issue can be resolved by padding the name attribute with zeros (leading zeros).

I don't understand why your modification does not work.
I modified source code and did generate, train and demo command.

**Comment by tgd-git** [comment]
Well it works here too, just tried again. My bad. It seems i have mixed up the model files when i tried yesterday (have trained the correct new one but used the old one when i run the demo) super facepalm! :)
Anyway, very happy we figured it out and it works now. Big thanks for your support!

**Comment by takuya-takeuchi** [comment]
Good.
OK, I will publish new nuget package after testing.

**Comment by takuya-takeuchi** [comment]
I released FaceRecignitionDotNet 1.2.3.14.
Have a try
